### ISSUE TITLE
Headings for items in guide pages

### ISSUE THREAD

**Issue by wbamberg** [mdn/stumptown-content#150]
Guide pages consist of chunks of prose with magic bits (live samples, BCD tables, possibly extra things in future) interspersed in them.

The main difference between guide pages and reference pages is that guide pages don't have any structure: the magic bits can be placed just anywhere, and the prose bits are not neat sections with H2 headings.

It would be nice if the renderer could render guide pages without any changes at all. We're pretty close to this. A sample guide page JSON is like:

```
{
  "title": "Applying color to HTML elements using CSS",
  "mdn_url": "https://developer.mozilla.org/docs/Web/HTML/Applying_color",
  "related_content": [
    ...
  ],
  "body": [
    {
      "type": "prose",
      "value": {
        "content": "<p>The use of color is a fundamental form of human expression..."
      }
    },
    {
      "type": "example",
      "value": {
        "description": {
          "title": "Specifying colors in stylesheets",
          "height": 192,
          "content": "<p>The easiest way..."
        },
        "sources": {
          "css": "....",
          "html": "..."
        }
      }
    },
    {
      "type": "prose",
      "value": {
        "content": "<h2>Letting the user pick a color</h2>..."
      }
    },
    {
      "type": "example",
      "value": {
        "description": {
          "title": "Picking a color",
          "height": 317,
          "content": "<p>..."
        },
        "sources": {
          "js": "...",
          "css": "...",
          "html": "..."
        }
      }
    },
    {
      "type": "prose",
      "value": {
        "content": "<h2>Using color wisely</h2>..."
      }
    }
  ]
}
```

As far as I can tell there are just a few things stopping the existing renderer from rendering this.

* titles are missing for prose sections: prose sections in reference pages (usually) get H2 headings ("Accessibility concerns" etc). But this isn't appropriate for prose sections in guide pages. This seems simple to deal with: [the renderer should just use `<Prose>` instead of `<ProseWithHeading>` if the `title` is omitted](https://github.com/mdn/stumptown-renderer/blob/master/client/src/document.js#L152).

* each live sample is represented in a property called `example`, while in reference pages live samples live in a property called `examples` containing an array of the things. But here we could just have guide pages match reference pages, so they represent a live sample with an `examples` property whose value is an array containing a single example.

* the rendering component for `examples` automatically adds an H2 "Examples": https://github.com/mdn/stumptown-renderer/blob/master/client/src/ingredients/examples.js#L88. But as with prose sections, this isn't appropriate for guide pages, since live samples can appear anywhere and aren't corralled into their own section of the page. What I think we should do here is:
    * in build-json, give the `examples` structure a `title: Examples` property when building reference pages, and omit when building guide pages
    * in the renderer, don't automatically add the H2 "Examples". Only, if `examples` in the JSON has `title`, use it for the H2, and if `title` is omitted from the JSON, omit the H2.

What do you think? @ddbeck , @Elchi3 , @peterbe, @joedarc  ?

**Comment by joedarc** [comment]
@wbamberg, i agree this should be pretty easily achievable. Already got a head start on stumptown-renderer rendering the guide pages just to see what all changes it would take for it to render nicely.

> the renderer should just use `<Prose>` instead of `<ProseWithHeading>` if the title is omitted

This would make it easy to differentiate when to use `<Prose>` vs `<ProseWithHeading>` for this scenario.

> But here we could just have guide pages match reference pages, so they represent a live sample with an examples property whose value is an array containing a single example.

I'm pretty sure that this would work because in the stumptown-renderer project i made an `<Example>` component that used almost identical code to render the content except it accepted a prop of a single example. 

> in build-json, give the examples structure a title: Examples property

Yes, again, this should make it easy to differentiate. 

With the proposed changes, it should require very little code in stumptown-renderer to work (mainly just check for existence of certain fields and render/not render stuff).

**Comment by ddbeck** [comment]
Yeah, this sounds sensible to me. A question on one thing:

> in build-json, give the `examples` structure a `title: Examples` property when building reference pages, and omit when building guide pages

Do I understand this correctly as reference pages get something like this:

```
    {
      "type": "example",
      "value": {
        "description": {
          "title": "Picking a color",
          "height": 317,
          "content": "<p>..."
        },
        …
      }
    }
```

while we strip out the title for guide pages?

```
    {
      "type": "example",
      "value": {
        "description": {
          "height": 317,
          "content": "<p>..."
        },
        …
      }
    }
```

If so, it struck me as a little surprising (consumers don't get all the "metadata" for an example everywhere). I don't think I'm opposed to it, really, but we might want to think about communicating to consumers that some of the atomic units of content (examples, BCD, whatever) are only complete or canonical via reference pages or something.

**Comment by joedarc** [comment]
@ddbeck unless i'm misunderstanding (I very well could be), it might be that instead of:

```
{
  "description": {
    "title": "Marking up an abbreviation semantically",
    "width": something,
    "height": something,
    "content": "<p>..."
  },
  "sources": {
    "html": "<p>..."
  }
}
```

for reference pages it could be:

```
{
  "title": "Examples",
  "description": {
    "title": "Marking up an abbreviation semantically",
    "width": something,
    "height": something,
    "content": "<p>..."
  },
  "sources": {
    "html": "<p>..."
  }
}
```

Where you keep the title in the description, but add an outer property of `title` to differentiate them.

**Comment by wbamberg** [comment]
Something like what @joedarc says. But having slept on it I'm not sure any more that this is a good idea. The `"examples"` in reference docs are a different kind of thing, semantically, from the `"example"` in guide pages. `"examples"` are a complete top-level section of the document, while an `"example"` is just a single live sample interpolated into the prose. So although it would be (a little) more code in the renderer to support both types, maybe it would be a better expression of what we actually mean.

**Comment by wbamberg** [comment]
Of course most of the code in the renderer is the code that actually renders each live sample, and that would be totally sharable.

**Comment by joedarc** [comment]
Off the top of my head i can think of two easy ways to get it working without changes to the json, let me know if you like either.

Could have an `<Example>` component along with `<Examples>` component and abstract out some of the reusable pieces?

Could modify the `<Examples>` component to accept a single example, if it's a singular one..don't render the `<h2>`

**Comment by joedarc** [comment]
If we end up agreeing that changes shouldn't be made to stumptown-content for rendering the guide pages we can probably carry on the conversation over there though.

**Comment by wbamberg** [comment]
> Could have an `<Example>` component along with `<Examples>` component and abstract out some of the reusable pieces?

I think this would be the way to go. If we end up hating it we can always reconsider.

> Could modify the `<Examples>` component to accept a single example, if it's a singular one..don't render the `<h2>`

There are "Examples" sections that only contain one example, so this would not work.

> we can probably carry on the conversation over there though

Yes :).

**Comment by peterbe** [comment]
Yes to everything! If it doesn't feel like you have answers to all the open questions you pose, it certainly feels like I agree with what you're leaning towards. 

Also, I just occurred to me that getting rid of the use of English here: https://github.com/mdn/stumptown-renderer/blob/e24684ac78689398463c482b38dac2e2d68fd98d/client/src/ingredients/examples.js#L88 would be great. The build-json should set (or not set) that and the renderer should:
```jsx
export function Examples({ title, examples }) {
  return (
    <>
      {title && <h2>{title}</h2>}
      {examples.map((example, i) => (
        <RenderExample key={i} example={example} />
      ))}
    </>
  );
}
```

Granted, this migh just move the problem from one space to another. Somewhere some French speaking translator needs to deal with this word and it feels like a `gettext` kind string rather than something that'd go into the content. I call this chrome English. 

However, a solution to that is to specify this thing in the front-matter on the docs that should have an `<h2>Examples</h2>`.

Here's what the front-matter looks like today:
```yaml
examples:
    - examples/simple-example
    - examples/multiple-sources-example
```
(from https://raw.githubusercontent.com/mdn/stumptown-content/master/content/html/reference/elements/video/docs.md)

What it could be is:
```yaml
examples:
    title: Examples
    sources:
      - examples/simple-example
      - examples/multiple-sources-example
```

Having to type `title: Examples` (or `title: Example`) for every single en-US reference page might be tedious. But explicit is almost always better than implicit and not a massive sit-up to have to do.
